elastic/logstash
# Support env variable in condition

## Release notes


Support environment variable in condition.

## Motivation

Two substitutions in ruby side cause inconsistent precedence when resolving the variable.
In the past, `${VAR:default_value}` is resolved in the following precedence: secret store -> environment variable -> default value.
After two substitutions, (first substitute)  environment variable -> default value -> (second substitute) secret store
The inconsistency could confuse users.

To truly support `if` with value in secret store, this commit implement substitution in java side 

## What does this PR do?



This PR substitutes ${VAR} in Expression, except RegexValueExpression, with the value in secret store, env. 
The substitution happens after syntax parsing and before graph execution.


## Why is it important/What is the impact to the user?



This PR allows users to use environment variable in condition, eg. `if [app] == "${VAR}"` 

## Checklist




## Author's Checklist


- checked expressions. `()`, `!()`, `>`, `>=`, `<`, `<=`, `==`, `!=`, `in`, `not in`, `=~`, `!~`, `and`, `or`, `nand`, `xor`
- regex `/${VAR}/` is not expected to do substitution
- checked multiple `and`. () and () and ()
- secret value deso not show in plain text when the pipeline config has syntax error
- undefined env variable show error msg when Logstash start 

## Known Limitation
Substitution only evaluates ${VAR} to string.
`if [app] == "${VAR}"` is a valid statement. VAR is resolved as a string.
`if [app] == 1` is a valid statement, but `if [app] == ${VAR}` is a wrong syntax. VAR cannot be evaluated to a number, hence cannot do number comparison.
This is the same behavior as of now.  The following count is a string and cannot be a number
```
generator {
  count => "${count}"
}
```
## How to test this PR locally


```
input {
	generator {
		lines => ['{"app": "foobar"}']
	}
}
filter {
	json {
		source => "message"
	}
	if [app] == "${VAR}" {
		mutate { add_tag => ["a_tag"] }
	}
}
output {
	stdout { codec => rubydebug }
}
```

run logstash with above config should see `tag` 
$ VAR=foobar bin/logstash 

## Related issues


Fixed: #5115
## Use cases



## Screenshots



## Logs
